crates/transaction-pool/src/pool/pending.rs

```diff
@@ -325,10 +347,63 @@ impl<T: TransactionOrdering> PendingPool<T> {
         id
     }
 
-    /// Removes the worst transaction from this pool.
-    pub(crate) fn pop_worst(&mut self) -> Option<Arc<ValidPoolTransaction<T::Transaction>>> {
-        let worst = self.all.iter().next().map(|tx| *tx.transaction.id())?;
-        self.remove_transaction(&worst)
+    /// Truncates the pool to the given limit.
+    ///
+    /// Removes transactions from the pending pool
+    /// The algorithm tries to reduce transaction counts by an approximately
+    /// equal number for all for accounts with many pending transactions.
+    pub(crate) fn truncate_pool(
+        &mut self,
+        limit: SubPoolLimit,
+    ) -> Vec<Arc<ValidPoolTransaction<T::Transaction>>> {
+        // return early if nothing needs to happen
+        if self.size() <= limit.max_size && self.len() <= limit.max_txs {
+            return Vec::new()
+        }
+
+        let mut removed = Vec::new();
+
+        // penalize non-local txs if limit is still exceeded
+        while self.size() > limit.max_size || self.len() > limit.max_txs {
```

Comment:
this is the inverse of L360 and L380?

I think we can rewrite this as `loop {}` and only keep a break